ext/curl/interface.c

```diff
@@ -2181,36 +2229,19 @@ static inline zend_result build_mime_structure_from_hash(php_curl *ch, zval *zpo
 			continue;
 		}
 
-		postval = zval_get_tmp_string(current, &tmp_postval);
+		if (Z_TYPE_P(current) == IS_ARRAY) {
+			zval *current_element;
+
+			ZEND_HASH_FOREACH_VAL(HASH_OF(current), current_element) {
+				add_simple_field(mime, string_key, current_element);
+			} ZEND_HASH_FOREACH_END();
 
-#if LIBCURL_VERSION_NUM >= 0x073800 /* 7.56.0 */
-		part = curl_mime_addpart(mime);
-		if (part == NULL) {
-			zend_tmp_string_release(tmp_postval);
 			zend_string_release_ex(string_key, 0);
-			return FAILURE;
-		}
-		if ((form_error = curl_mime_name(part, ZSTR_VAL(string_key))) != CURLE_OK
-			|| (form_error = curl_mime_data(part, ZSTR_VAL(postval), ZSTR_LEN(postval))) != CURLE_OK) {
-			error = form_error;
+			continue;
 		}
-#else
-		/* The arguments after _NAMELENGTH and _CONTENTSLENGTH
-			* must be explicitly cast to long in curl_formadd
-			* use since curl needs a long not an int. */
-		form_error = curl_formadd(&first, &last,
-								CURLFORM_COPYNAME, ZSTR_VAL(string_key),
-								CURLFORM_NAMELENGTH, ZSTR_LEN(string_key),
-								CURLFORM_COPYCONTENTS, ZSTR_VAL(postval),
-								CURLFORM_CONTENTSLENGTH, ZSTR_LEN(postval),
-								CURLFORM_END);
 
-		if (form_error != CURL_FORMADD_OK) {
-			/* Not nice to convert between enums but we only have place for one error type */
-			error = (CURLcode)form_error;
-		}
-#endif
-		zend_tmp_string_release(tmp_postval);
+		add_simple_field(mime, string_key, current);
```

Comment:
Nit: An else-case and removing the `continue` could remove the duplicate `zend_string_release_ex`.

Reply:
Yes, but IMO the current code is more readable, as each special block now looks the same. If we'd to add one in the future, then it's less likely the wrong thing gets copied. (I'm also allergic to `else`).